Index: synapse/synapse/_scripts/register_new_matrix_user.py
===================================================================
--- synapse.orig/synapse/_scripts/register_new_matrix_user.py
+++ synapse/synapse/_scripts/register_new_matrix_user.py
@@ -42,6 +42,8 @@ No 'registration_shared_secret' or 'regi
 
 _DEFAULT_SERVER_URL = "http://localhost:8008"
 
+class UserExistsError(RuntimeError):
+    pass
 
 def request_registration(
     user: str,
@@ -59,7 +61,7 @@ def request_registration(
     r = requests.get(url)
 
     if r.status_code != 200:
-        _print("ERROR! Received %d %s" % (r.status_code, r.reason))
+        _print("ERROR! Received %d %s (%s)" % (r.status_code, r.reason, r.json()))
         if 400 <= r.status_code < 500:
             try:
                 _print(r.json()["error"])
@@ -103,6 +105,9 @@ def request_registration(
                 _print(r.json()["error"])
             except Exception:
                 pass
+
+            if r.json()['errcode'] == 'M_USER_IN_USE':
+                raise UserExistsError()
         return exit(1)
 
     _print("Success!")
@@ -217,6 +222,12 @@ def main() -> None:
     )
 
     parser.add_argument(
+        "--idempotent",
+        action="store_true",
+        help="Command will not fail if user already exists.",
+    )
+
+    parser.add_argument(
         "server_url",
         nargs="?",
         help="URL to use to talk to the homeserver. By default, tries to find a "
@@ -269,9 +280,16 @@ def main() -> None:
     if args.admin or args.no_admin:
         admin = args.admin
 
-    register_new_user(
-        args.user, args.password, server_url, secret, admin, args.user_type
-    )
+    try:
+        register_new_user(
+            args.user, args.password, server_url, secret, admin, args.user_type
+        )
+    except UserExistsError:
+        if args.idempotent:
+            print('Ignoring the fact that the user already exists.')
+            pass
+        else:
+            sys.exit(1)
 
 
 def _read_file(file_path: Any, config_path: str) -> str:
