Index: synapse/synapse/storage/databases/main/room.py
===================================================================
--- synapse.orig/synapse/storage/databases/main/room.py
+++ synapse/synapse/storage/databases/main/room.py
@@ -20,6 +20,7 @@
 #
 #
 
+import json
 import logging
 from abc import abstractmethod
 from enum import Enum
@@ -514,6 +515,26 @@ class RoomWorkerStore(CacheInvalidationW
                 search_term.lower(),
             ]
 
+        complex_where_clauses = []
+        complex_where_clauses_type: str = None
+        if search_filter and search_filter.get(
+            PublicRoomsFilterFields.COMPLEX_SEARCH_TERMS, None
+        ):
+            complex_where_clauses_type = search_filter[PublicRoomsFilterFields.COMPLEX_SEARCH_TERMS]['type']
+            for search_term in search_filter[PublicRoomsFilterFields.COMPLEX_SEARCH_TERMS]['terms']:
+                # only allow searching for these specific fields
+                for field in set(search_term.keys()) & {'canonical_alias', 'name', 'topic'}:
+                    complex_where_clauses.append(
+                        f"""
+                        (
+                            LOWER({field}) LIKE ?
+                        )
+                        """
+                    )
+                    query_args += [
+                        search_term[field].lower(),
+                    ]
+
         room_type_clause, args = self._construct_room_type_where_clause(
             search_filter.get(PublicRoomsFilterFields.ROOM_TYPES, None)
             if search_filter
@@ -527,6 +548,9 @@ class RoomWorkerStore(CacheInvalidationW
         if where_clauses:
             where_clause = " AND " + " AND ".join(where_clauses)
 
+        if complex_where_clauses:
+            where_clause += " AND ( " + f" {complex_where_clauses_type} ".join(complex_where_clauses) + " )"
+
         dir = "DESC" if forwards else "ASC"
         sql = f"""
             SELECT
Index: synapse/synapse/api/constants.py
===================================================================
--- synapse.orig/synapse/api/constants.py
+++ synapse/synapse/api/constants.py
@@ -285,6 +285,7 @@ class PublicRoomsFilterFields:
     """
 
     GENERIC_SEARCH_TERM: Final = "generic_search_term"
+    COMPLEX_SEARCH_TERMS: Final = "complex_search_terms"
     ROOM_TYPES: Final = "room_types"
 
 
